src/spaces/walk.jl

```diff
@@ -109,14 +126,31 @@ function randomwalk!(
         ))
     end
     offsets = offsets_at_radius(model, r)
-    walk!(agent, rand(abmrng(model), offsets), model; kwargs...)
+    if ifempty && force_motion
+        n_attempts = 2*length(offsets)
+        while n_attempts != 0
```

Comment:
it is (much) more performant often since it doesn't require to collect all positions if there at least an available position, `random_agent` is done with the same logic

Reply:
actually there was also another problem, I should have used `move_agent!` not `walk!` inside the loop :Q.
Still I plan to benchmark how much we gain to do so

Reply:
> it is more performant often since it doesn't require to collect all positions if there at least an available position, `random_agent` is done with the same logic. It will be slower though if there will be no empty positions

"Often" depends on the simulation. If I'm running a simulation with high packing then it will be often inefficient.
I guess the optimal solution would be to rewrite explicitly a loop identical to `sampling_with_condition_single` that evaluates whether an offset is accessible only inside the while loop, without the need to collect anything at the start.

Reply:
there is still a concern though I like your methodology but it's not that easy to do what is done in `sampling_with_condition_single` in this case, it would require 1 of this two things since otherwise we will make less efficient memory management for other subsequent operations:
1. copy `offsets`
2. sort `offsets` after the while loop

I think 1. is preferable but we need some benchmark to see how much we loose in the other sense, when there is at least one available position, and then decide.

Reply:
wait a second, you make me wonder one maybe very important thing: what about using SVector instead of Vector for `offsets`?!

Reply:
Why? Maybe I miss something but there should be no intrinsic order in `offsets`, I don't see a problem with shuffling its elements around. I believe `offsets` is always used through `rand` anyway so should be alright.

Reply:
the fact is that `offsets` is used to find nearby positions/agents, look at how the nearby stuff are implemented to see this, so keeping the order is better, see [Write cache-friendly codes](https://julialang.org/blog/2013/09/fast-numeric/#write_cache-friendly_codes). So all in all I think what you suggested will make the code less performant except when we have no positions available, so I'm kinda inclined to merge the PR as it is, and come back to this matter later :-)